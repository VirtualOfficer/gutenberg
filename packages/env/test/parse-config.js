'use strict';
/* eslint-disable jest/no-conditional-expect */
/**
 * External dependencies
 */
const path = require( 'path' );
const { readFile } = require( 'fs' ).promises;

/**
 * Internal dependencies
 */
const { loadConfig } = require( '../lib/config' );
const detectDirectoryType = require( '../lib/config/detect-directory-type' );

jest.mock( 'fs', () => ( {
	promises: {
		readFile: jest.fn(),
		stat: jest.fn().mockResolvedValue( true ),
	},
} ) );

// This mocks a small response with a format matching the stable-check API.
// It makes getLatestWordPressVersion resolve to "100.0.0".
jest.mock( 'got', () =>
	jest.fn( ( url ) => ( {
		json: () => {
			if ( url === 'https://api.wordpress.org/core/stable-check/1.0/' ) {
				return Promise.resolve( {
					'1.0': 'insecure',
					'99.1.1': 'outdated',
					'100.0.0': 'latest',
					'100.0.1': 'fancy',
				} );
			}
		},
	} ) )
);

jest.mock( '../lib/config/detect-directory-type', () => jest.fn() );

describe( 'Config Parsing', () => {
	beforeEach( () => {
		process.env.WP_ENV_HOME = '/cache';
		detectDirectoryType.mockResolvedValue( null );
	} );

	afterEach( () => {
		delete process.env.WP_ENV_HOME;
	} );

	it( 'should use default configuration', async () => {
		readFile.mockImplementation( async () => {
			throw { code: 'ENOENT' };
		} );

		const config = await loadConfig( '/test/gutenberg' );

		expect( config ).toMatchSnapshot();
	} );

	it( 'should load local configuration', async () => {
		readFile.mockImplementation( async ( fileName ) => {
			if ( fileName === '/test/gutenberg/.wp-env.json' ) {
				return JSON.stringify( {
					core: 'WordPress/WordPress#trunk',
					port: 123,
				} );
			}

			throw { code: 'ENOENT' };
		} );

		const config = await loadConfig( path.resolve( '.' ) );

		expect( config ).toMatchSnapshot();
	} );

	it( 'should load local and override configurations', async () => {
		readFile.mockImplementation( async ( fileName ) => {
			if ( fileName === '/test/gutenberg/.wp-env.json' ) {
				return JSON.stringify( {
					core: 'WordPress/WordPress#trunk',
					port: 123,
					testsPort: 456,
				} );
			}

			if ( fileName === '/test/gutenberg/.wp-env.override.json' ) {
				return JSON.stringify( {
					port: 999,
				} );
			}

			throw { code: 'ENOENT' };
		} );

		const config = await loadConfig( path.resolve( '/test/gutenberg' ) );

		expect( config ).toMatchSnapshot();
	} );
} );
/* eslint-enable jest/no-conditional-expect */
