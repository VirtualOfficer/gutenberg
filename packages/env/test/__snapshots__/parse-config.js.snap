// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Parsing should load local and override configurations 1`] = `
{
  "configDirectoryPath": "/test/gutenberg",
  "detectedLocalConfig": true,
  "dockerComposeConfigPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/docker-compose.yml",
  "env": {
    "development": {
      "config": {
        "SCRIPT_DEBUG": true,
        "WP_DEBUG": true,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:999",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:999",
        "WP_TESTS_DOMAIN": "localhost:999",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "path": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "ref": "trunk",
        "testsPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 999,
      "themeSources": [],
    },
    "tests": {
      "config": {
        "SCRIPT_DEBUG": false,
        "WP_DEBUG": false,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:456",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:456",
        "WP_TESTS_DOMAIN": "localhost:456",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "path": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "ref": "trunk",
        "testsPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 456,
      "themeSources": [],
    },
  },
  "name": "gutenberg",
  "workDirectoryPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338",
}
`;

exports[`Config Parsing should load local configuration 1`] = `
{
  "configDirectoryPath": "/Users/obliviousharmony/Code/WordPress/gutenberg",
  "detectedLocalConfig": true,
  "dockerComposeConfigPath": "/cache/f406b227b400575d350730477365a884/docker-compose.yml",
  "env": {
    "development": {
      "config": {
        "SCRIPT_DEBUG": true,
        "WP_DEBUG": true,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:8888",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:8888",
        "WP_TESTS_DOMAIN": "localhost:8888",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/f406b227b400575d350730477365a884/WordPress",
        "path": "/cache/f406b227b400575d350730477365a884/WordPress",
        "ref": "100.0.0",
        "testsPath": "/cache/f406b227b400575d350730477365a884/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 8888,
      "themeSources": [],
    },
    "tests": {
      "config": {
        "SCRIPT_DEBUG": false,
        "WP_DEBUG": false,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:8889",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:8889",
        "WP_TESTS_DOMAIN": "localhost:8889",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/f406b227b400575d350730477365a884/WordPress",
        "path": "/cache/f406b227b400575d350730477365a884/WordPress",
        "ref": "100.0.0",
        "testsPath": "/cache/f406b227b400575d350730477365a884/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 8889,
      "themeSources": [],
    },
  },
  "name": "gutenberg",
  "workDirectoryPath": "/cache/f406b227b400575d350730477365a884",
}
`;

exports[`Config Parsing should use default configuration 1`] = `
{
  "configDirectoryPath": "/test/gutenberg",
  "detectedLocalConfig": true,
  "dockerComposeConfigPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/docker-compose.yml",
  "env": {
    "development": {
      "config": {
        "SCRIPT_DEBUG": true,
        "WP_DEBUG": true,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:8888",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:8888",
        "WP_TESTS_DOMAIN": "localhost:8888",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "path": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "ref": "100.0.0",
        "testsPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 8888,
      "themeSources": [],
    },
    "tests": {
      "config": {
        "SCRIPT_DEBUG": false,
        "WP_DEBUG": false,
        "WP_ENVIRONMENT_TYPE": "local",
        "WP_HOME": "http://localhost:8889",
        "WP_PHP_BINARY": "php",
        "WP_SITEURL": "http://localhost:8889",
        "WP_TESTS_DOMAIN": "localhost:8889",
        "WP_TESTS_EMAIL": "admin@example.org",
        "WP_TESTS_TITLE": "Test Blog",
      },
      "coreSource": {
        "basename": "WordPress",
        "clonePath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "path": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/WordPress",
        "ref": "100.0.0",
        "testsPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338/tests-WordPress",
        "type": "git",
        "url": "https://github.com/WordPress/WordPress.git",
      },
      "mappings": {},
      "phpVersion": null,
      "pluginSources": [],
      "port": 8889,
      "themeSources": [],
    },
  },
  "name": "gutenberg",
  "workDirectoryPath": "/cache/5fea4c5689ef6cc4a4e6eaaa39323338",
}
`;
